pkg/odo/cli/registry/add.go

```diff
@@ -0,0 +1,86 @@
+package registry
+
+import (
+	// Build-in packages
+	"fmt"
+
+	// Third-party packages
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	ktemplates "k8s.io/kubernetes/pkg/kubectl/util/templates"
+
+	// odo packages
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	"github.com/openshift/odo/pkg/preference"
+)
+
+const addCommandName = "add"
+
+// "odo registry add" command description and examples
+var (
+	addLongDesc = ktemplates.LongDesc(`Add devfile registry`)
+
+	addExample = ktemplates.Examples(`# Add devfile registry
+	%[1]s cheregistry https://che-devfile-registry.openshift.io/
+	`)
+)
+
+// AddOptions encapsulates the options for the "odo registry add" command
+type AddOptions struct {
+	operation    string
+	registryName string
+	registryURL  string
+}
+
+// NewAddOptions creates a new AddOptions instance
+func NewAddOptions() *AddOptions {
+	return &AddOptions{}
+}
+
+// Complete completes AddOptions after they've been created
+func (o *AddOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	o.operation = "add"
+	o.registryName = args[0]
+	o.registryURL = args[1]
+	return
+}
+
+// Validate validates the AddOptions based on completed values
+func (o *AddOptions) Validate() (err error) {
+	return
```

Comment:
Should we also have a validation that the user's trying to add a valid registry?
```sh
$ odo registry list
NAME       URL
myreg2     example.com

$ odo catalog list components
 ⚠  Registry myreg2 is not set up properly with error: Unable to download the devfile index.json from example.com/devfiles/index.json: Get example.com/devfiles/index.json: unsupported protocol scheme ""
 ✗  no deployable components found
```

Reply:
> Should we also add validation to ensure that user is not mistakenly adding a registry with different name but same URL. For example, I could manage to do something like below:
> 
> ```shell
> $ odo registry list
> NAME                       URL
> CheDevfileRegistry         https://che-devfile-registry.openshift.io/
> DefaultDevfileRegistry     https://raw.githubusercontent.com/elsony/devfile-registry/master
> myreg                      https://che-devfile-registry.openshift.io/
> myreg2                     https://che-devfile-registry.openshift.io/
> ```

This could be a perfectly valid use-case. It should be possible to have the same registry URL under different names

Reply:
> Should we also have a validation that the user's trying to add a valid registry?
> 
> ```shell
> $ odo registry list
> NAME       URL
> myreg2     example.com
> 
> $ odo catalog list components
>  ⚠  Registry myreg2 is not set up properly with error: Unable to download the devfile index.json from example.com/devfiles/index.json: Get example.com/devfiles/index.json: unsupported protocol scheme ""
>  ✗  no deployable components found
> ```

Early validation in `odo registry add` has also its downsides. It means that if there is a temporary network error, or for example, you want a registry that is inside the corporate network and you need VPN to access it you won't be able to do it unless it is accessible in a time of running odo registry add. Not saying that this is a necessarily bad thing, but just something to consider

`⚠  Registry myregistry is not set up properly... ` is currently a warning that doesn't make the whole common fail. This is a good thing 👍 We don't want to make odo fail when there is a connection issue with just one registry that might not even be used in the command.


```
▶ odo registry add myregistry fff
New registry was successfully added

▶ odo registry list
NAME                       URL
CheDevfileRegistry         https://che-devfile-registry.openshift.io/
DefaultDevfileRegistry     https://raw.githubusercontent.com/elsony/devfile-registry/master
myregistry                 fff

▶ odo catalog list components
 ⚠  Registry myregistry is not set up properly with error: Unable to download the devfile index.json from fff/devfiles/index.json: Get fff/devfiles/index.json: unsupported protocol scheme ""
Odo OpenShift Components:
NAME        PROJECT       TAGS                                        SUPPORTED
nodejs      openshift     10,8,8-RHOAR,latest                         YES
dotnet      openshift     2.0,latest                                  NO
httpd       openshift     2.4,latest                                  NO
nginx       openshift     1.10,1.12,1.8,latest                        NO
nodejs      openshift     6                                           NO
perl        openshift     5.24,5.26,latest                            NO
php         openshift     7.0,7.1,latest                              NO
python      openshift     2.7,3.5,3.6,latest                          NO
ruby        openshift     2.3,2.4,2.5,latest                          NO
wildfly     openshift     10.0,10.1,11.0,12.0,13.0,8.1,9.0,latest     NO

Odo Devfile Components:
NAME                 DESCRIPTION                           REGISTRY                   SUPPORTED
maven                Upstream Maven and OpenJDK 11         DefaultDevfileRegistry     YES
nodejs               Stack with NodeJS 10                  DefaultDevfileRegistry     YES
openLiberty          Open Liberty microservice in Java     DefaultDevfileRegistry     YES
java-spring-boot     Spring Boot® using Java               DefaultDevfileRegistry     Y
```

Reply:
@dharmit We do have validation logic to validate user specify registry, basically we validate the following things:
- Whether the registry URL itself is reachable or not
- Whether the index file format is correct or not as the registry URL points to the index file
- Whether the devfile format is correct or not as the index file points to devfile

But we don't do the early validation as Tomas mentions above, we validate when users **use** those registries (eg. odo create, odo catalog), it would efficient we can do the registry validation with getting values from registry together so that we can save one generic "iteration" from time complexity perspective.

Reply:
@kadel 
> `⚠ Registry myregistry is not set up properly... ` is currently a warning that doesn't make the whole common fail. This is a good thing +1 We don't want to make odo fail when there is a connection issue with just one registry that might not even be used in the command.

@GeekArthur 
> But we don't do the early validation as Tomas mentions above, we validate when users use those registries (eg. odo create, odo catalog), it would efficient we can do the registry validation with getting values from registry together so that we can save one generic "iteration" from time complexity perspective.

Makes sense to me. :+1: